postings/pilosa/codec.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pilosa
+
+import (
+	"bytes"
+
+	"github.com/m3db/m3ninx/postings"
+
+	"github.com/pilosa/pilosa/roaring"
+)
+
+// Encoder helps serialize a Pilosa RoaringBitmap
+type Encoder struct {
+	scratchBuffer bytes.Buffer
+}
+
+// NewEncoder returns a new Encoder.
+func NewEncoder() *Encoder {
+	return &Encoder{}
+}
+
+// Reset resets the internal state of the encoder to allow
+// for re-use.
+func (e *Encoder) Reset() {
+	e.scratchBuffer.Reset()
+}
+
+// Encode encodes the provided postings list.
+func (e *Encoder) Encode(pl postings.List) ([]byte, error) {
+	e.scratchBuffer.Reset()
+
+	pilosaBitmap, err := toPilosa(pl)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = pilosaBitmap.WriteTo(&e.scratchBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	return e.scratchBuffer.Bytes(), nil
```

Comment:
Would we need to return a copy of these bytes since we reuse the buffer?

Reply:
Can push the responsibility on the caller. Added a note about sematics to the method comment.

Reply:
What about having the caller provide the buffer so it's explicit that they control the lifetime of the buffer?